ispcrt/ispcrt.cpp

```diff
@@ -78,6 +87,408 @@ static OBJECT_T &referenceFromHandle(HANDLE_T handle) {
+#endif
+
+// OS agnostic function to dynamically load a shared library.
+void *dyn_load_lib(const char *name, const char *name_major_version, const char *name_full_version) {
+#if defined(_WIN32) || defined(_WIN64)
+    return LoadLibrary(name);
+#else
+    // Try to load a device library starting from the most specific name down to more general one.
+    void *handle = dlopen(name_full_version, RTLD_NOW | RTLD_LOCAL);
+    if (handle) {
+        return handle;
+    }
+    handle = dlopen(name_major_version, RTLD_NOW | RTLD_LOCAL);
+    if (handle) {
+        return handle;
+    }
+    // This is the most general name, e.g., libispcrt_device_cpu.so. Under some
+    // circumstances, it may points to older or newer version. This is probably
+    // not a big deal until incompatible changes are encountered between versions.
+    return dlopen(name, RTLD_NOW | RTLD_LOCAL);
+#endif
+}
+
+// OS agnostic function to get an address of symbol from the previously loaded shared library.
+void *dyn_load_sym(void *handle, const char *symbol) {
+#if defined(_WIN32) || defined(_WIN64)
+    return (void*) GetProcAddress((HMODULE)handle, symbol);
+#else
+    return dlsym(handle, symbol);
+#endif
+}
+
+
+// CPU device API.
+typedef void (*ISPCLaunchF)(void**, void*, void*, int, int, int);
+typedef void* (*ISPCAllocF)(void**, int64_t, int32_t);
+typedef void (*ISPCSyncF)(void*);
+static ISPCLaunchF ispc_launch_fptr = nullptr;
+static ISPCAllocF ispc_alloc_fptr = nullptr;
+static ISPCSyncF ispc_sync_fptr = nullptr;
+
+#ifndef ISPCRT_BUILD_STATIC
+// During static linking this API is delivered by linking the real implementation.
+// ispc expects these functions to have C linkage / not be mangled
+// Stubs to load and call actual implementations (*_cpu) from detail/cpu/ispc_tasking.cpp.
+extern "C" {
+void ISPCLaunch(void **handlePtr, void *f, void *data, int countx, int county, int countz) {
+    if (ispc_launch_fptr) {
+        return ispc_launch_fptr(handlePtr, f, data, countx, county, countz);
+    }
+    // TODO: this code can be called directly from ISPC code. Not sure how to exit error safely.
+    throw std::runtime_error("Missing ISPCLaunch symbol");
+}
+
+void *ISPCAlloc(void **handlePtr, int64_t size, int32_t alignment) {
+    if (ispc_alloc_fptr) {
+        return ispc_alloc_fptr(handlePtr, size, alignment);
+    }
+    // TODO: this code can be called directly from ISPC code. Not sure how to exit error safely.
+    throw std::runtime_error("Missing ISPCAlloc symbol");
+}
+
+void ISPCSync(void *handle) {
+    if (ispc_sync_fptr) {
+        return ispc_sync_fptr(handle);
+    }
+    // TODO: this code can be called directly from ISPC code. Not sure how to exit error safely.
+    throw std::runtime_error("Missing ISPCSync symbol");
+}
+}
+#endif
+
+// Auxiliary function to load device CPU solib and initialize tasking API pointers if build with tasking support.
+void *handleCPUDevilceLib() {
```

Comment:
Typo?

Reply:
I guess you mean `handleCPUDevilceLib ` => `handleCPUDeviceLib `, right?

Reply:
Yes, it is. I have fixed that.